### ISSUE TITLE
Flaky test: DateHeaderValueManagerTests

### ISSUE THREAD

**Issue by JunTaoLuo** [aspnet/KestrelHttpServer#1049]
I'm seeing this test fail in our CI: http://aspnetci/viewLog.html?buildId=126592&buildTypeId=Lite_UniverseTest&tab=buildLog#_

```
[Microsoft.AspNetCore.Server.KestrelTests.DateHeaderValueManagerTests.GetDateHeaderValue_ReturnsUpdatedValueAfterIdle] Assert.Equal() Failure
                                   (pos 23)
Expected: úúú, 16 Aug 2016 22:51:28 GMT
Actual:   úúú, 16 Aug 2016 22:51:18 GMT
                                   (pos 23)
   at Microsoft.AspNetCore.Server.KestrelTests.DateHeaderValueManagerTests.<GetDateHeaderValue_ReturnsUpdatedValueAfterIdle>d__2.MoveNext() in C:\BuildAgent\work\33bdfc1cae7b2a38\KestrelHttpServer\test\Microsoft.AspNetCore.Server.KestrelTests\DateHeaderValueManagerTests.cs:line 101
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

This is inconsistent but still would warrant some investigation.

**Comment by JunTaoLuo** [comment]
http://aspnetci/viewLog.html?buildId=126746&tab=buildResultsDiv&buildTypeId=XPlat_Mac_MacUniverseCoreCLR
Also seeing this flaky test on mac
Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.RegisterAddresses_IPv4_Success
```
[Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.RegisterAddresses_IPv4_Success(addressInput: "http://*:8001/;https://*:8002/", testUrls: Func`2 { Method = System.String[] <get_AddressRegistrationDataIPv4>b__4(Microsoft.AspNetCore.Hosting.Server.Features.IServerAddressesFeature), Target = <>c__DisplayClass13_0 { port1 = 8001, port2 = 8002 } })] System.IO.IOException : Authentication failed because the remote party has closed the transport stream.
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Testing.HttpClientSlim.<GetStream>d__6.MoveNext() in /Users/asplab/buildAgent/work/3d42ba28d4e3c9/KestrelHttpServer/test/shared/HttpClientSlim.cs:line 92
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Testing.HttpClientSlim.<GetStringAsync>d__1.MoveNext() in /Users/asplab/buildAgent/work/3d42ba28d4e3c9/KestrelHttpServer/test/shared/HttpClientSlim.cs:line 24
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.<RegisterAddresses_Success>d__7.MoveNext() in /Users/asplab/buildAgent/work/3d42ba28d4e3c9/KestrelHttpServer/test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/AddressRegistrationTests.cs:line 91
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.<RegisterAddresses_IPv4_Success>d__0.MoveNext() in /Users/asplab/buildAgent/work/3d42ba28d4e3c9/KestrelHttpServer/test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/AddressRegistrationTests.cs:line 28
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by JunTaoLuo** [comment]
More flaky tests

HttpsTests.ClientHandshakeFailureLoggedAsInformation
http://aspnetci/project.html?projectId=XPlat&testNameId=-7731905546387681977&tab=testDetails

HttpsTests.EmptyRequestLoggedAsInformation
http://aspnetci/project.html?projectId=XPlat&testNameId=-3947247458576941802&tab=testDetails

**Comment by halter73** [comment]
#1071 should fix the AddressRegistrationTests issue on OS X. @cesarbs is looking into a better fix that will disable parallel test execution only on os x due to the ulimit issues.

We should also look into "BadHttpRequestTests.TestInvalidHeaders(rawHeaders: "Header-1: value1\r\nHeader-2: value2\r\n") " whic looks to be regressed by https://github.com/aspnet/KestrelHttpServer/commit/8836eec7d886fcbcb918c43fa20b66bcc8199874

**Comment by cesarbs** [comment]
@JunTaoLuo https://github.com/aspnet/KestrelHttpServer/commit/84efe624a627e767f08d5a6f263bd40bd7e80fed fixes the last two you reported.

**Comment by JunTaoLuo** [comment]
@cesarbs nice! So this task is done then?

**Comment by cesarbs** [comment]
Are you still seeing the flakiness on `DateHeaderValueManagerTests`?

**Comment by JunTaoLuo** [comment]
Seems like it's been passing consistently for a week now.

**Comment by cesarbs** [comment]
Ok, closing this then.

**Comment by halter73** [comment]
I expect DateHeaderValueManagerTests will fail again. We can reopen when it does though.

**Comment by natemcmaster** [comment]
Failed again this morning.
```
Assert.Equal() Failure
                                 ↓ (pos 23)
Expected: ···, 30 Mar 2017 04:31:14 GMT
Actual:   ···, 30 Mar 2017 04:31:04 GMT
                                 ↑ (pos 23)
   at Microsoft.AspNetCore.Server.KestrelTests.DateHeaderValueManagerTests.<GetDateHeaderValue_ReturnsUpdatedValueAfterIdle>d__3.MoveNext() in C:\BuildAgent\work\33bdfc1cae7b2a38\.r\KestrelHttpServer\test\Microsoft.AspNetCore.Server.KestrelTests\DateHeaderValueManagerTests.cs:line 108
```

**Comment by mikeharder** [comment]
Do we understand the root cause of this failure?  Test bug or product bug?

**Comment by halter73** [comment]
I took a look. This is almost certainly a test bug. Given the right conditions, I think the timer (which is scheduled to run every 100 ms) isn't firing before an `await Task.Delay(TimeSpan.FromSeconds(1));` completes.

I'm not sure what the best fix is short of changing the timer interval or delay.

**Comment by mikeharder** [comment]
One way would be for `DateHeaderValueManager` to take an `ITimer` interface which can be mocked, like `ISystemClock`.  Otherwise I agree, there's no way to make a reliable test if the product code depends on a `Timer` firing.

**Comment by mikeharder** [comment]
Or change the test to wait longer if the expected value is not seen after 1 second.  Check after one second, and retry say 5 times if the first attempt didn't match.  I'm not sure what the test is really trying to verify.

**Comment by mikeharder** [comment]
Another failure: http://aspnetci/viewLog.html?buildId=200519&tab=buildResultsDiv&buildTypeId=Lite_UniverseTest

**Comment by halter73** [comment]
Fixed by #1649
